src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb

```diff
@@ -1647,17 +1647,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End SyncLock
 
             If completedCompilationUnit Then
-                EventQueue.Enqueue(New CompilationUnitCompletedEvent(Me, tree))
+                EventQueue.TryEnqueue(New CompilationUnitCompletedEvent(Me, tree))
             End If
 
             If completedCompilation Then
-                EventQueue.Enqueue(New CompilationCompletedEvent(Me))
-                EventQueue.Complete() ' signal the End Of compilation events
+                EventQueue.TryEnqueue(New CompilationCompletedEvent(Me))
+                EventQueue.PromiseNotToEnqueue()
+                EventQueue.TryComplete() ' signal the End Of compilation events
             End If
         End Sub
 
         Friend Function ShouldAddEvent(symbol As Symbol) As Boolean
-            If EventQueue Is Nothing Then
+            If EventQueue Is Nothing OrElse EventQueue.IsCompleted Then
```

Comment:
The structure of the compiler is such that it produces all events before promising not to enqueue any longer. Once we promise not to enqueue any longer, a flag is set inside the async queue and the queue throws an exception if any event is queued. In other words, we do check, but at a different layer. Those changes are all included in this PR.